### DISCUSSION TITLE
Adding /Enforce viewer preference

### DISCUSSION THREAD

**Discussion by samuelbradshaw** [py-pdf/pypdf#2509] [Q&A] [answered]
In the PDF 2.0 specification, there's a way to prevent a PDF from being scaled to anything other than 100% when printing (at least in Acrobat; other PDF viewers don't consistently respect it). This is done by by adding these two viewer preferences to the PDF:

/Enforce [/PrintScaling]
/PrintScaling /None

I was able to set `/PrintScaling /None` with `writer.viewer_preferences.print_scaling = '/None'`, but I haven't figured out how to set `/Enforce [/PrintScaling]`. Is this possible? The closest I've gotten is `/Enforce (PrintScaling)`. Here's my code:

```
from pypdf import PdfReader, PdfWriter
from pypdf.generic import NameObject, TextStringObject

reader = PdfReader('input.pdf')
writer = PdfWriter()

for page in reader.pages:
  writer.add_page(page)

writer.create_viewer_preferences()

# Add viewer preference: /Enforce [/PrintScaling]
# The code below doesn't work – it gives me /Enforce (PrintScaling)
vp = writer._root_object['/ViewerPreferences'].get_object()
vp.update({ NameObject('/Enforce'): TextStringObject('PrintScaling') })

# Add viewer preference: /PrintScaling /None
writer.viewer_preferences.print_scaling = '/None'

with open('output.pdf', 'wb') as f:
  writer.write(f)
```

See [PDF 2.0 specification, page 430](https://developer.adobe.com/document-services/docs/assets/5b15559b96303194340b99820d3a70fa/PDF_ISO_32000-2.pdf#page=436)

**Comment by pubpub-zz** [comment]
This feature has been introduced in PDF 2.0, this is why it has not been implemented yet. I just note a small error : 
`vp.update({ NameObject('/Enforce'): ArrayObject([TextStringObject('PrintScaling')])})`

can you confirm it is working?

I should be able to propose a PR

**Reply by samuelbradshaw** [reply]
Thank you! It didn't work:

`vp.update({ NameObject('/Enforce'): ArrayObject([TextStringObject('PrintScaling')]) })`
Produces this:
`/Enforce [ (PrintScaling) ]`
(I checked in Acrobat and Acrobat didn't recognize it as valid)

`vp.update({ NameObject('/Enforce'): ArrayObject([TextStringObject('/PrintScaling')]) })`
Produces this:
`/Enforce [ (\057PrintScaling) ]`
(Acrobat didn't recognize this either)

`vp.update({ NameObject('/Enforce'): ArrayObject(TextStringObject('/PrintScaling')) })`
`vp.update({ NameObject('/Enforce'): ArrayObject('/PrintScaling') })`
`vp.update({ NameObject('/Enforce'): ArrayObject(['/PrintScaling']) })`
These give me an error: AttributeError: 'str' object has no attribute 'write_to_stream'

This is what I'm hoping to get:
`/Enforce [/PrintScaling]`
(I've verified that this works in Acrobat – tested by manually editing the PDF in a text editor)

**Reply by pubpub-zz** [reply, marked as answer]
oups  I missed the error:
`vp.update({ NameObject('/Enforce'): ArrayObject([NameObject('/PrintScaling')])})`

**Reply by samuelbradshaw** [reply]
Yes, that worked! Thank you!! :)

**Reply by pubpub-zz** [reply]
PR #2511 is under review to ease this
